llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1450,9 +1460,24 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
           NewMul->setHasNoSignedWrap(true);
           return NewMul;
         }
+
+        // Special case: lshr nuw (mul (X, 3), 1) -> add nuw nsw (X, lshr(X, 1)
+        if (ShAmtC == 1 && MulC->getZExtValue() == 3) {
```

Comment:
Is there a more general form for this than these specific constants? Why is this the one special case?

Reply:
Can you add an alive2 link?

Reply:
> Is there a more general form for this than these specific constants? Why is this the one special case? 

(X+(x<<1))>>1 is canonicalized to x*3/2 so this is missed as a result

Reply:
> Can you add an alive2 link? 

Pretty sure I did in the first comment I made